src/analyses/invariant_propagation.h

```diff
@@ -48,6 +42,8 @@ class invariant_propagationt:public
   typedef ait<invariant_set_domaint> baset;
 
 protected:
+  friend invariant_set_domain_factoryt;
```

Comment:
Each `invariant_set_domain` needs access to a few of the fields of the invariant_propagation object.  This is a kinda wonky encapsulation violation but it would require a substantial refactor to remove this.  Instead what I have done is the `invariant_set_domain_factory` is created with a reference to the `invariant_propagationt` and creates the domain objects with references to the relevant parts of `invariant_propagationt`.  To do this it has to be a friend.

It's ugly but what it replaced was worse.

Reply:
Ok, adding your insights above as comment in the code might help future refactoring efforts.

Reply:
Added as a comment.